datafusion/common/src/scalar/mod.rs

```diff
@@ -3343,6 +3339,14 @@ impl ScalarValue {
         arr1 == &right
     }
 
+    /// Compare `self` with `other` and return an `Ordering`.
+    /// Returns `Err` if the values cannot be compared, e.g., they have incompatible data types.
+    pub fn try_cmp(&self, other: &Self) -> Result<Ordering> {
+        self.partial_cmp(other).ok_or_else(|| {
+            _internal_datafusion_err!("Uncomparable values: {self:?}, {other:?}")
```

Comment:
Is this an internal error, in the "it arises from a bug in DataFusion" sense?

If it can  come from a malformed query / user input I think it might be better to return a `DataFusionError::Execution` instead.